contracts/Stake/StakeManager.sol

```diff
@@ -152,286 +153,258 @@ contract StakeManager is Initializable, Ownable {
         }
-        stashes[stashId] = Stash(msg.sender, address(0), 0, _tokens);
-        emit StashCreated(msg.sender, stashId, stashIndex, _tokens, _amounts);
-        return stashId;
+        stashes[_stashId].staker = msg.sender;
+        emit StashCreated(msg.sender, _stashId, _stashIndex, _tokens, _amounts);
+        stashIndex = _stashIndex + 1;  // Can't overflow
+        return _stashId;
     }
 
     function addToStash(
-        bytes32 _stashId, 
-        bytes32[] memory _tokens, 
+        bytes32 _stashId,
+        bytes32[] memory _tokens,
         uint256[] memory _amounts
     ) public {
-        Stash memory stash = stashes[_stashId];
+        Stash memory _stash = stashes[_stashId];
         require(
-            stash.staker == msg.sender, 
+            _stash.staker == msg.sender,
             "StakeManager:addToStash - Only staker can delegate stash to a cluster"
         );
         require(
-            stash.undelegatesAt <= block.number,
+            _stash.undelegatesAt <= block.number,
             "StakeManager:addToStash - Can't add to stash during undelegation"
         );
         require(
-            _tokens.length == _amounts.length, 
+            _tokens.length == _amounts.length,
             "StakeManager:addToStash - Each tokenId should have a corresponding amount and vice versa"
         );
-        if(stash.delegatedCluster != address(0)) {
-            rewardDelegators.delegate(msg.sender, stash.delegatedCluster, _tokens, _amounts);
+        if(_stash.delegatedCluster != address(0)) {
+            rewardDelegators.delegate(msg.sender, _stash.delegatedCluster, _tokens, _amounts);
         }
-        uint256 index = stashes[_stashId].tokensDelegated.length;
-        for(uint256 i=0; i < _tokens.length; i++) {
+        for(uint256 i = 0; i < _tokens.length; i++) {
+            bytes32 _tokenId = _tokens[i];
             require(
-                tokenAddresses[_tokens[i]].isActive, 
+                tokenAddresses[_tokenId].isActive,
                 "StakeManager:addToStash - Invalid tokenId"
             );
             if(_amounts[i] != 0) {
-                TokenData memory tokenData = stashes[_stashId].amount[_tokens[i]];
-                if(tokenData.amount == 0) {
-                    stashes[_stashId].tokensDelegated.push(_tokens[i]);
-                    stashes[_stashId].amount[_tokens[i]] = TokenData(_amounts[i], index);
-                    index++;
-                } else {
-                    stashes[_stashId].amount[_tokens[i]].amount = tokenData.amount.add(_amounts[i]);
-                }
-                _lockTokens(_tokens[i], _amounts[i], msg.sender);
+                stashes[_stashId].amount[_tokenId] = stashes[_stashId].amount[_tokenId].add(_amounts[i]);
+                _lockTokens(_tokenId, _amounts[i], msg.sender);
             }
         }
         // TODO: If gas usage for emitting tokens and amount is high, then query using txHash and remove them from event
```

Comment:
Querying using graph becomes hard if we have to query from tx. So let this stay.